tests/storages_tests/test_storages.py

```diff
@@ -352,336 +452,406 @@ def test_get_trial_number_from_id(storage_mode):
+        assert storage.get_trial(trial_id_1).intermediate_values == {0: 0.3, 2: 0.4}
+        assert storage.get_trial(trial_id_2).intermediate_values == {}
+        assert storage.get_trial(trial_id_3).intermediate_values == {0: 0.1, 1: 0.4, 2: 0.5}
 
-    storage = storage_init_func()
-    trial_id_1 = storage.create_new_trial(storage.create_new_study())
+        # Test setting existing step.
+        assert not storage.set_trial_intermediate_value(trial_id_1, 0, 0.3)
 
-    def check_set_and_get(trial_id, key, value):
-        # type: (int, str, Any) -> None
+        non_existent_trial_id = max(trial_id_1, trial_id_2, trial_id_3) + 1
+        with pytest.raises(KeyError):
+            storage.set_trial_intermediate_value(non_existent_trial_id, 0, 0.2)
 
-        storage.set_trial_user_attr(trial_id, key, value)
-        assert storage.get_trial(trial_id).user_attrs[key] == value
+        storage.set_trial_state(trial_id_1, TrialState.COMPLETE)
+        # Cannot change values of finished trials.
+        with pytest.raises(RuntimeError):
+            storage.set_trial_intermediate_value(trial_id_1, 0, 0.2)
+
+
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES)
+def test_get_trial_user_attrs(storage_mode: str) -> None:
 
-    # Test setting value.
-    for key, value in EXAMPLE_ATTRS.items():
-        check_set_and_get(trial_id_1, key, value)
-    assert storage.get_trial(trial_id_1).user_attrs == EXAMPLE_ATTRS
+    with StorageSupplier(storage_mode) as storage:
+        _, study_to_trials = _setup_studies(storage, n_study=2, n_trial=5, seed=10)
+        assert all(
+            storage.get_trial_user_attrs(trial_id) == trial.user_attrs
+            for trials in study_to_trials.values()
+            for trial_id, trial in trials.items()
+        )
 
-    # Test overwriting value.
-    check_set_and_get(trial_id_1, "dataset", "ImageNet")
+        non_existent_trial = max(tid for ts in study_to_trials.values() for tid in ts.keys()) + 1
+        with pytest.raises(KeyError):
+            storage.get_trial_user_attrs(non_existent_trial)
 
-    # Test another trial.
-    trial_id_2 = storage.create_new_trial(storage.create_new_study())
-    check_set_and_get(trial_id_2, "baseline_score", 0.001)
-    assert len(storage.get_trial(trial_id_2).user_attrs) == 1
-    assert storage.get_trial(trial_id_2).user_attrs["baseline_score"] == 0.001
 
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES)
+def test_set_trial_user_attr(storage_mode: str) -> None:
 
-@parametrize_storage
-def test_set_and_get_trial_system_attr(storage_init_func):
-    # type: (Callable[[], BaseStorage]) -> None
+    with StorageSupplier(storage_mode) as storage:
+        trial_id_1 = storage.create_new_trial(storage.create_new_study())
 
-    storage = storage_init_func()
-    study_id = storage.create_new_study()
-    trial_id_1 = storage.create_new_trial(study_id)
+        def check_set_and_get(trial_id, key, value):
+            # type: (int, str, Any) -> None
```

Comment:
How about using Python3-style type hinting?
```suggestion
        def check_set_and_get(trial_id: int, key: str, value: Any) -> None:
```